client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,444 @@
+	for blockHeight := currentTip.height; blockHeight > lastScannedBlock.height; blockHeight-- {
+		blockHash, err := dcr.node.GetBlockHash(dcr.ctx, blockHeight)
+		if err != nil {
+			return false, translateRPCCancelErr(err)
+		}
+		blockFilter, err := dcr.getBlockFilterV2(blockHash)
+		if err != nil {
+			return false, err
+		}
+		if !blockFilter.MatchAny(txScripts) {
+			continue // check previous block's filters (blockHeight--)
+		}
+		dcr.log.Debugf("Block %d (%s) likely contains tx %s. Confirming.", blockHeight, blockHash, tx.hash)
+		blk, err := dcr.getBlock(blockHash, true)
+		if err != nil {
+			return false, err
+		}
+		blockTxs := append(blk.RawTx, blk.RawSTx...)
+		for i := range blockTxs {
+			blkTx := &blockTxs[i]
+			if blkTx.Txid == tx.hash.String() {
+				dcr.log.Debugf("Found mined tx %s in block %d (%s).", tx.hash, blockHeight, blockHash)
+
+				msgTx, err := msgTxFromHex(blkTx.Hex)
+				if err != nil {
+					return false, fmt.Errorf("invalid hex for tx %s: %v", tx.hash, err)
+				}
+
+				tx.relevantBlock = &block{hash: blockHash, height: blockHeight}
+				tx.tree = determineTxTree(msgTx)
+
+				// Store the pkScripts for all of this tx's outputs so they
+				// can be used to later to determine if an output is spent.
+				tx.outputs = make(map[uint32]*externalTxOutput)
+				for i := range blkTx.Vout {
+					output := &blkTx.Vout[i]
+					amt, err := dcrutil.NewAmount(output.Value)
+					if err != nil {
+						dcr.log.Errorf("tx output %s:%d has invalid amount: %v", tx.hash, i, err)
+						continue
+					}
+					pkScript, err := hex.DecodeString(output.ScriptPubKey.Hex)
+					if err != nil {
+						dcr.log.Errorf("tx output %s:%d has invalid pkScript: %v", tx.hash, i, err)
+						continue
+					}
+					tx.outputs[uint32(i)] = &externalTxOutput{
+						// TODO: output.ScriptPubKey.Version with dcrd 1.7 *release*, not yet
+						TxOut:  newTxOut(int64(amt), output.Version, pkScript),
+						txHash: tx.hash,
+						vout:   uint32(i),
+					}
+				}
+				return true, nil
+			}
+		}
+	}
+
+	// Scan completed from current tip to last scanned block. Set the
+	// current tip as the last scanned block so subsequent scans cover
+	// the latest tip back to this current tip.
+	tx.lastScannedBlock = currentTip.hash
+	dcr.log.Debugf("Tx %s NOT found in blocks %d (%s) to %d (%s).", tx.hash,
+		currentTip.height, currentTip.hash, lastScannedBlock.height, lastScannedBlock.hash)
+	return false, nil
+}
+
+// findExternalTxOutputSpender returns true if a block is found that contains
+// a tx that spends the provided output AND the block is a mainchain block.
+// If no block has been found to contain a tx that spends the provided output
+// or the block that was found to contain the spender is re-orged out of the
+// mainchain, this method matches the output's pkScript against multiple block
+// cfilters (starting from the block containing the output up till the current
```

Comment:
To determine the purpose of `txBlockHash` at a glance, perhaps a minor doc revision: "...starting from the block containing the output, **txBlockHash**, up to the current..."

And then perhaps something about using `lastScannedBlock` if a previous scan was already performed, up to current...